### ISSUE TITLE
Concurrent modification exception in websock connector

### ISSUE THREAD

**Issue by mchitre** [org-arl/fjage#112]
```
Exception in thread "OutputThread:ws://127.0.1.1:80/ws" java.util.ConcurrentModificationException
at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
at java.util.ArrayList$Itr.next(ArrayList.java:859)
at org.arl.fjage.connectors.WebSocketConnector$OutputThread.run(WebSocketConnector.java:174)
```
Observed in release `1.6.3`

**Comment by mchitre** [comment]
Strange because the `wsHandlers` list is protected by a `Collections.synchronizedList()`, and the iterator is protected by `synchronized(wsHandlers) { ... }`.

**Comment by notthetup** [comment]
Seen again when the `fjage.js` receive timeout error occurs.

```
Exception in thread "OutputThread:ws://127.0.1.1:80/ws" java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
	at java.util.ArrayList$Itr.next(ArrayList.java:859)
	at org.arl.fjage.connectors.WebSocketConnector$OutputThread.run(WebSocketConnector.java:174)```

**Comment by notthetup** [comment]
Do we need to move from the `for` loop to `while(iterator.hasNext())`?

https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedList-java.util.List-

**Comment by notthetup** [comment]
Nah. It seems it'll be the same.

https://stackoverflow.com/questions/7335958/java-synchronized-list-for-loop

**Comment by mchitre** [comment]
For the record: This seems to be the cause for `fjage.js` failing after network disconnection at times, while the shell websocket seems to be fine.

**Comment by mchitre** [comment]
I think I understand why this happens! Crazy!!

We have:
```
        synchronized(wsHandlers) {
          for (WSHandler t: wsHandlers)
            t.write(s);
        }
```
The list access is always synchronized, and this loop doesn't modify the list (supposedly), so we should expect everything to be fine! But, the `t.write()` calls Jetty's `session.getRemote().sendString(s)`, and that probably calls `onClose()` from the same thread, if the socket is dead. And in `onClose()`, we have: `wsHandlers.remove(this)`. Since we are in the same thread, the lock is already acquired, and so the `remove()` happens while iterating!!!!

**Comment by mchitre** [comment]
Fixed in `dev` for 1.7 by using `CopyOnWriteArrayList` that allows concurrent modifications.

@notthetup do you want to pick it up in `1.6.5`? Also @prasadtiru @AnneThKu @notthetup ,would be nice to see if we can try to reproduce the problem with this fix to see if we have really solved it.

**Comment by mchitre** [comment]
Closing it. Reopen if problem persists.
